src/client-side-encryption/providers/index.ts

```diff
@@ -0,0 +1,169 @@
+   */
+  local?: {
+    /**
+     * The master key used to encrypt/decrypt data keys.
+     * A 96-byte long Buffer or base64 encoded string.
+     */
+    key: Buffer | string;
+  };
+
+  /**
+   * Configuration options for using 'kmip' as your KMS provider
+   */
+  kmip?: {
+    /**
+     * The output endpoint string.
+     * The endpoint consists of a hostname and port separated by a colon.
+     * E.g. "example.com:123". A port is always present.
+     */
+    endpoint?: string;
+  };
+
+  /**
+   * Configuration options for using 'azure' as your KMS provider
+   */
+  azure?:
+    | {
+        /**
+         * The tenant ID identifies the organization for the account
+         */
+        tenantId: string;
+
+        /**
+         * The client ID to authenticate a registered application
+         */
+        clientId: string;
+
+        /**
+         * The client secret to authenticate a registered application
+         */
+        clientSecret: string;
+
+        /**
+         * If present, a host with optional port. E.g. "example.com" or "example.com:443".
+         * This is optional, and only needed if customer is using a non-commercial Azure instance
+         * (e.g. a government or China account, which use different URLs).
+         * Defaults to "login.microsoftonline.com"
+         */
+        identityPlatformEndpoint?: string | undefined;
+      }
+    | {
+        /**
+         * If present, an access token to authenticate with Azure.
+         */
+        accessToken: string;
+      }
+    | Record<string, never>;
+
+  /**
+   * Configuration options for using 'gcp' as your KMS provider
+   */
+  gcp?:
+    | {
+        /**
+         * The service account email to authenticate
+         */
+        email: string;
+
+        /**
+         * A PKCS#8 encrypted key. This can either be a base64 string or a binary representation
+         */
+        privateKey: string | Buffer;
+
+        /**
+         * If present, a host with optional port. E.g. "example.com" or "example.com:443".
+         * Defaults to "oauth2.googleapis.com"
+         */
+        endpoint?: string | undefined;
+      }
+    | {
+        /**
+         * If present, an access token to authenticate with GCP.
+         */
+        accessToken: string;
+      }
+    | Record<string, never>;
+}
+
+/**
+ * Auto credential fetching should only occur when the provider is defined on the kmsProviders map
+ * and the settings are an empty object.
+ *
+ * This is distinct from a nullish provider key.
+ *
+ * @internal - exposed for testing purposes only
+ */
+export function isEmptyCredentials(provider: KMSProvider, kmsProviders: KMSProviders) {
+  return (
+    provider in kmsProviders &&
+    kmsProviders[provider] != null &&
+    typeof kmsProviders[provider] === 'object' &&
+    // Typescript does not infer that kmsProviders[provider] is non-null,
+    // even though we check it two lines up.
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
I do not feel strongly, rather than explain this to the next reader/dev, I would just pull `kmsProviders[provider]` out to a variable and narrow the variable.